sdks/java/harness/src/main/java/org/apache/beam/fn/harness/WindowMergingFnRunner.java

```diff
@@ -130,16 +130,16 @@ private MergingViaWindowFnRunner(WindowFn<?, W> windowFn) {
       this.mergedWindows = new ArrayList<>();
       this.currentWindows = new ArrayList<>();
       this.mergeContext =
-          windowFn.new MergeContext() {
+          ((WindowFn) windowFn).new MergeContext() {
```

Comment:
Is it not the case that this is a `WindowFn<T, W>`?

Reply:
It's a `WindowFn<?, W>`, and anonymous classes whose enclosing instance types contain captured wildcards are illegal: https://bugs.openjdk.java.net/browse/JDK-8203436.

Reply:
I meant whether it was, in fact, legitimate to treat it as `WindowFn<T, W>`. I believe the intent of the wildcard is probably to acknowledge that the type is unknown. So changing `?` to `Object` would probably work fine and have fewer casts and no rawtypes.

Sorry for the nitpicking, just trying to avoid rawtypes. IMO only for compatibility with pre-generics code or extremely special cases where Java's type system can't handle it.

Reply:
Thanks for clarifying, using `WindowFn<T, W>` within `MergingViaWindowFnRunner` also works around the breakage. I updated the PR to do that instead.